### PR TITLE
Cover a few edge cases in SatInt better

### PR DESCRIPTION
**Pull request by michaelpj**
Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-core/plutus-core/src/Data/SatInt.hs [hunk 1]
```diff
@@ -214,32 +214,42 @@ instance Integral SatInt where
...
-     | a == minBound && b == (-1) = minBound
+     -- a/-1 = -a, -minBound = maxBound
+     -- We can't just fall though since `divInt` would overflow instead
+     | a == minBound && b == (-1) = maxBound
      | otherwise                  = SI (a `divInt` b)
 
     SI a `mod` SI b
      | b == 0                     = divZeroError
-     | a == minBound && b == (-1) = minBound
+     -- a/-1 = -a, exactly
```
Comment (kwxm) [inline plutus-core/plutus-core/src/Data/SatInt.hs:238]: "exactly" meaning "with no remainder"?

**Review by kwxm — APPROVED** [review]
I really hate this div/mod/quot/rem stuff! I think I've managed to convince myself that this is right though.

File: plutus-core/plutus-core/satint-test/TestSatInt.hs [hunk 2]
```diff
@@ -53,7 +53,9 @@ tests =
     unitTest "min-"    (saturatesNeg ((minBound :: SatInt) - 1)),
     unitTest "1/0"     (isArithException (1 `div` 0)),
     unitTest "min*-1"  (saturatesPos ((minBound :: SatInt) * (-1))),
-    unitTest "min/-1"  (saturatesNeg ((minBound :: SatInt) `div` (-1))),
+    unitTest "0-min"   (saturatesPos (0 - (minBound :: SatInt))),
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-core/plutus-core/satint-test/TestSatInt.hs:59]: I experimented with the property tests a bit. First of all, they weren't using a very good generator, so I fixed that, which led to lots of div by zero errors (because it wasn't picking 0 before!), so I fixed the handling for that. Then I broke the implementation again, but even with ratcheting up the number of tests, it didn't pick `minBound` and `-1` ever. Eventually I wrote a manual generator that prefers the edge cases and then I managed to make it break. In the process I found out that `gcd` and `lcm` are fundamentally broken though, which is good!

File: plutus-core/plutus-core/src/Data/SatInt.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-core/plutus-core/src/Data/SatInt.hs:238]: yeah

File: plutus-core/plutus-core/satint-test/TestSatInt.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kwxm) [reply plutus-core/plutus-core/satint-test/TestSatInt.hs:59]: >  Eventually I wrote a manual generator that prefers the edge cases

I had to do something very similar with the tests I wrote for the `upperIntegerQuotient` function a couple of days ago.  I used Hedgehog since that's what was already being used for the rest of the tests in `plutus-core` but the default generators didn't seem to do a very good job.  There's a `Size` parameter involved that changes hoe much of a range is covered, but the documentation doesn't really describe how that's used (I think maybe it's involved shrinking).  Experimentation showed that if you were asking for samples from [a,b] it would only choose numbers from the lower 30% with the default `Size` value.  Eventually I had to write special generators for both ends and the middle of the `SatInt` range.  I suppose the moral is "check your generators".
